src/github.com/mozilla-services/pushgo/simplepush/handlers_endpoint_test.go

```diff
@@ -0,0 +1,606 @@
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+
+	"github.com/rafrombrc/gomock/gomock"
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+type keyTest struct {
+	name     string
+	key      string
+	expected bool
+}
+
+func (t keyTest) Test() error {
+	actual := validPK(t.key)
+	if actual != t.expected {
+		return fmt.Errorf("On test %s, got %s; want %s", t.name, actual,
+			t.expected)
+	}
+	return nil
+}
+
+func formReader(vals url.Values) io.ReadCloser {
+	return ioutil.NopCloser(strings.NewReader(vals.Encode()))
+}
+
+func getJSON(header http.Header, body *bytes.Buffer) (*bytes.Buffer, bool) {
+	if header.Get("Content-Type") != "application/json" {
+		return nil, false
+	}
+	buf := new(bytes.Buffer)
+	if err := json.Compact(buf, body.Bytes()); err != nil {
+		return nil, false
+	}
+	return buf, true
+}
+
+func TestEndpointValidKey(t *testing.T) {
+	tests := []keyTest{
+		{"Hyphenated key", "bb817e78-c568-4780-90c3-471f3b74f055.29f72d42-7a2a-42e4-897b-dc2c8a7cb676", true},
+		{"Non-hyphenated key", "bb817e78c568478090c3471f3b74f055.29f72d427a2a42e4897bdc2c8a7cb676", true},
+		{"Non-hyphenated, case-insensitive key", "BB817E78C568478090C3471F3B74F055.29F72D427A2A42E4897BDC2C8A7CB676", true},
+		{"Valid Base64-encoded key", "u4F+eMVoR4CQw0cfO3TwVQ==.KfctQnoqQuSJe9wsiny2dg==", false},
+		{"Invalid UUID; valid key", "pQrS-tUv-1234.wXyZ-5678", true},
+		{"Non-alphanumeric characters", "\t\r\n", false},
+		{"Extraneous punctuators", "_=!@#$%^&*()[]", false},
+	}
+	for _, test := range tests {
+		if err := test.Test(); err != nil {
+			t.Error(err)
+		}
+	}
+}
+
+func TestEndpointResolveKey(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mckLogger := NewMockLogger(mockCtrl)
+	mckLogger.EXPECT().ShouldLog(gomock.Any()).Return(true).AnyTimes()
+	mckLogger.EXPECT().Log(gomock.Any(), gomock.Any(),
+		gomock.Any(), gomock.Any()).AnyTimes()
+
+	mckStat := NewMockStatistician(mockCtrl)
+	mckStore := NewMockStore(mockCtrl)
+
+	Convey("Endpoint tokens", t, func() {
+		app := NewApplication()
+		app.SetLogger(mckLogger)
+		app.SetMetrics(mckStat)
+		app.SetStore(mckStore)
+
+		Convey("Should return a 404 for invalid tokens", func() {
+			app.SetTokenKey("c3v0AlmmxXu_LSfdZY3l3eayLsIwkX48")
+			eh := NewEndpointHandler()
+			eh.setApp(app)
+			app.SetEndpointHandler(eh)
+
+			resp := httptest.NewRecorder()
+			req := &http.Request{
+				Method: "PUT",
+				Header: http.Header{},
+				URL: &url.URL{
+					Path: "/update/j1bqzFq9WiwFZbqay-y7xVlfSvtO1eY="}, // "123.456"
+			}
+			gomock.InOrder(
+				mckStore.EXPECT().KeyToIDs("123.456").Return("", "", false),
+				mckStat.EXPECT().Increment("updates.appserver.invalid"),
+			)
+			eh.ServeMux().ServeHTTP(resp, req)
+
+			So(resp.Code, ShouldEqual, 404)
+			body, isJSON := getJSON(resp.HeaderMap, resp.Body)
+			So(isJSON, ShouldBeTrue)
+			So(body.String(), ShouldEqual, `"Invalid Token"`)
+		})
+
+		Convey("Should not decode plaintext tokens", func() {
```

Comment:
Nit: Should not decode plaintext tokens *if there's no encryption key specified* (otherwise, it should reject them.)